File: src/context.cpp
#include <cstring>
#include <cinttypes>
#include <optional>
#include <stdexcept>
#include <string>
#include <stdio.h>
#include <fstream>
#include <stdarg.h>
#include "context.hpp"
#include "util.hpp"

const char *VALID_ARGS[] = {
    "-f", "-i", "-o", "-m", "-a", "-mem", "-harts", "-hz", "-cL", "-c1",
    "-c2", "-c3", "-cS1", "-cS2", "-cS3", "-cB1", "-cB2", "-cB3", "-cW1", "-cW2",
    "-cW3", "-cE1", "-cE2", "-cE3", "-cC", "-bp", "-bpD", "-mmu", "-tlb", "-tlbE"
};
const int NUM_VALID_ARGS = sizeof(VALID_ARGS) / sizeof(char *);

bool Context::is_digit(char src) {
    return src >= '0' && src <= '9';
}

bool Context::is_letter(char src) {
    return (src >= 'a' && src <= 'z') || (src >= 'A' && src <= 'Z');
}

char Context::to_uppercase(char src) {
    if (src >= 'a' && src <= 'z') {
        return src - 'a' + 'A';
    }
    return src;
}

CacheCoherency Context::parse_cache_coherency(std::string value, std::string name) {
    if (!strcmp(value.c_str(), "none")) {
        return NONE;
    }
    else if (!strcmp(value.c_str(), "snoop")) {
        return SNOOP;
    }
    else if (!strcmp(value.c_str(), "dir")) {
        return DIR;
    }

    char msg[100];
    sprintf(msg, "Invalid argument for \"%s\": \"%s\"", name.c_str(), value.c_str());
    throw std::invalid_argument(msg);
}

EvictionPolicy Context::parse_eviction_policy(std::string value, std::string name) {
    if (!strcmp(value.c_str(), "fifo")) {
        return FIFO;
    }
    else if (!strcmp(value.c_str(), "lru")) {
        return LRU;
    }
    else if (!strcmp(value.c_str(), "lfu")) {
        return LFU;
    }

    char msg[100];
    sprintf(msg, "Invalid argument for \"%s\": \"%s\"", name.c_str(), value.c_str());
    throw std::invalid_argument(msg);
}

std::pair<CacheAssociativity, int> Context::parse_cache_type(std::string value, std::string name) {
    if (value.size() < 2LU) {
        char msg[100];
        sprintf(msg, "Invalid argument for \"%s\": \"%s\"", name.c_str(), value.c_str());
        throw std::invalid_argument(msg);
    }

    const char start[] = {value[0], value[1], '\0'};
    if (!strcmp(start, "dm")) {
        return std::pair<CacheAssociativity, int>(DIRECT_MAPPED, -1);
    }
    else if (!strcmp(start, "fa")) {
        return std::pair<CacheAssociativity, int>(FULL_ASSOC, -1);
    }
    else if (!strcmp(start, "sa")) {
        int ways;
        if (!sscanf(value.c_str(), "sa,%d", &ways)) {
            char msg[100];
            sprintf(msg, "Failed to parse numeric for \"%s\": \"%s\"", name.c_str(), value.c_str());
            throw std::invalid_argument(msg);
        }
        return std::pair<CacheAssociativity, int>(SET_ASSOC, ways);
    }

    char msg[100];
    sprintf(msg, "Invalid argument for \"%s\": \"%s\"", name.c_str(), value.c_str());
    throw std::invalid_argument(msg);
}

int Context::parse_int(std::string value, std::string name) {
    int output;
    if (!sscanf(value.c_str(), "%d", &output)) {
        char msg[100];
        sprintf(msg, "Failed to parse argument for \"%s\": \"%s\"", name.c_str(), value.c_str());
        throw std::invalid_argument(msg);
    }
    return output;
}

uint64_t Context::parse_metric(std::string value, std::string name) {
    // confirm every non-terminal char is digit
    for (uint64_t i = 0; i < value.size() - 1; ++i) {
        if (!is_digit(value[i])) {
            char msg[100];
            sprintf(msg, "Invalid numeric for argument \"%s\": \"%s\".", name.c_str(), value.c_str());
            throw std::invalid_argument(msg);
        }
    }

    // confirm terminal char is digit or letter
    char end = value[value.size() - 1];
    if (!is_digit(end) && !is_letter(end)) {
        char msg[100];
        sprintf(msg, "Invalid value for argument \"%s\": \"%s\".", name.c_str(), value.c_str());
        throw std::invalid_argument(msg);
    }

    // handle only digits
    uint64_t output;
    if (is_digit(end)) {
        sscanf(value.c_str(), "%lu", &output);
        return output;
    }

    // handle letter at end
    // - validate letter
    if (end != 'K' && end != 'M' && end != 'G') {
        char msg[100];
        sprintf(msg, "Invalid metric prefix for argument \"%s\": \"%c\".", name.c_str(), end);
        throw std::invalid_argument(msg);
    }

    // - parse numeric
    uint64_t numeric;
    sscanf(value.substr(0, value.size() - 1).c_str(), "%lu", &numeric);

    // - multiply by corresponding power of two
    switch (end) {
        case 'K':
            return numeric * 1'024;
        case 'M':
            return numeric * 1'048'576;
        case 'G':
            return numeric * 1'073'741'824;
    }

    char msg[100];
    sprintf(msg, "Error parsing metric value for argument \"%s\": \"%s\".", name.c_str(), value.c_str());
    throw std::invalid_argument(msg);
}

bool Context::parse_bool(std::string value, std::string name, const char *truthy, const char *falsey) {
    if (!strcmp(value.c_str(), truthy)) return true;
    else if (!strcmp(value.c_str(), falsey)) return false;
    char msg[100];
    sprintf(msg, "Argument for \"%s\" must be %s\\%s, not \"%s\".", name.c_str(), truthy, falsey, value.c_str());
    throw std::invalid_argument(msg);
}

void Context::read_command_line(std::vector<std::string> &vargs) {
    for (uint64_t i = 1; i < vargs.size(); i += 2) {
        const char *varg = vargs.at(i).c_str();

        // confirm argument is valid
        bool is_valid_arg = false;
        for (int j = 0; j < NUM_VALID_ARGS; ++j) {
            if (!strcmp(varg, VALID_ARGS[j])) {
                is_valid_arg = true;
                break;
            }
        }
        if (!is_valid_arg) {
            char msg[100];
            sprintf(msg, "Invalid argument: \"%s\".", varg);
            throw std::invalid_argument(msg);
        }

        // confirm space for additional argument
        if (i == vargs.size() - 1) {
            char msg[100];
            sprintf(msg, "Missing value for argument: \"%s\".", varg);
            throw std::invalid_argument(msg);
        }
        std::string next = vargs.at(i + 1);

        // unholy arg handling of dispair :(
        if (strcmp(varg, "-f") == 0) {
            config_file = next;
        }
        else if (strcmp(varg, "-i") == 0) {
            elf_file = next;
        }
        else if (strcmp(varg, "-o") == 0) {
            output_file = next;
        }
        else if (strcmp(varg, "-m") == 0) {
            allow_rv32m = parse_bool(next, varg, "on", "off");
        }
        else if (strcmp(varg, "-a") == 0) {
            allow_rv32a = parse_bool(next, varg, "on", "off");
        }
        else if (strcmp(varg, "-mem") == 0) {
            ram_size = parse_metric(next, "mem");
        }
        else if (strcmp(varg, "-harts") == 0) {
            num_harts = parse_int(next, "harts");
        }
        else if (strcmp(varg, "-hz") == 0) {
            cycle_frequency = parse_metric(next, "hz");
        }
        else if (strcmp(varg, "-cL") == 0) {
            cache_depth = parse_int(next, "cL");
        }
        else if (strcmp(varg, "-c1") == 0) {
            std::pair<CacheAssociativity, int> parsed = parse_cache_type(next, "c1");
            l1_cache.ca = parsed.first;
            l1_cache.ways = parsed.second;
        }
        else if (strcmp(varg, "-c2") == 0) {
            std::pair<CacheAssociativity, int> parsed = parse_cache_type(next, "c2");
            l2_cache.ca = parsed.first;
            l2_cache.ways = parsed.second;
        }
        else if (strcmp(varg, "-c3") == 0) {
            std::pair<CacheAssociativity, int> parsed = parse_cache_type(next, "c3");
            l3_cache.ca = parsed.first;
            l3_cache.ways = parsed.second;
        }
        else if (strcmp(varg, "-cS1") == 0) {
            l1_cache.size = parse_metric(next, "cS1");
        }
        else if (strcmp(varg, "-cS2") == 0) {
            l2_cache.size = parse_metric(next, "cS2");
        }
        else if (strcmp(varg, "-cS3") == 0) {
            l3_cache.size = parse_metric(next, "cS3");
        }
        else if (strcmp(varg, "-cB1") == 0) {
            l1_cache.block_size = parse_int(next, "cB1");
        }
        else if (strcmp(varg, "-cB2") == 0) {
            l2_cache.block_size = parse_int(next, "cB2");
        }
        else if (strcmp(varg, "-cB3") == 0) {
            l3_cache.block_size = parse_int(next, "cB3");
        }
        else if (strcmp(varg, "-cW1") == 0) {
            l1_cache.is_write_back = parse_bool(next, "cW1", "wb", "wt");
        }
        else if (strcmp(varg, "-cW2") == 0) {
            l2_cache.is_write_back = parse_bool(next, "cW2", "wb", "wt");
        }
        else if (strcmp(varg, "-cW3") == 0) {
            l3_cache.is_write_back = parse_bool(next, "cW3", "wb", "wt");
        }
        else if (strcmp(varg, "-cE1") == 0) {
            l1_cache.eviction = parse_eviction_policy(next, "cE1");
        }
        else if (strcmp(varg, "-cE2") == 0) {
            l2_cache.eviction = parse_eviction_policy(next, "cE2");
        }
        else if (strcmp(varg, "-cE3") == 0) {
            l3_cache.eviction = parse_eviction_policy(next, "cE3");
        }
        else if (strcmp(varg, "-cC") == 0) {
            cache_coherency = parse_cache_coherency(next, "cC");
        }
        else if (strcmp(varg, "-bp") == 0) {
            branch_prediction_rows = parse_int(next, "bp");
        }
        else if (strcmp(varg, "-bpD") == 0) {
            default_prediction = parse_int(next, "bpD");
        }
        else if (strcmp(varg, "-mmu") == 0) {
            allow_mmu = parse_bool(vargs.at(i + 1), varg, "on", "off");
        }
        else if (strcmp(varg, "-tlb") == 0) {
            num_tlb_slots = parse_int(next, "tlb");
        }
        else if (strcmp(varg, "-tlbE") == 0) {
            tlb_eviction = parse_eviction_policy(next, "tlbE");
        }
        else {
            // should never be reached
            char msg[100];
            sprintf(msg, "Invalid argument: \"%s\".", varg);
            throw std::invalid_argument(msg);
        }
    }
}

std::optional<std::string> Context::get_config_file(std::vector<std::string> &vargs) {
    std::optional<std::string> output = std::nullopt;
    for (uint64_t i = 0; i < vargs.size(); ++i) {
        if (!std::strcmp("-f", vargs.at(i).c_str())) {
            if (i != vargs.size() - 1) {
                output = std::string(vargs.at(i+1).c_str());   
            } else {
                throw std::invalid_argument("Missing file name for configuration file.");
            }
        }
    }
    return output;
}

void Context::read_config_file(std::string file_path) {
    std::ifstream file(file_path);
    if (!file) {
        char msg[100];
        sprintf(msg, "Failed to access file: \"%s\".", file_path.c_str());
        throw std::runtime_error(msg);
    }
    
    // process file into fake command-line arguments
    std::vector<std::string> vargs;
    vargs.push_back("fake argv[0]");
    std::string line;
    while (std::getline(file, line)) {
        // ignore comments and strip
        const uint64_t ignore_idx = line.find("#");
        if (ignore_idx != std::string::npos) {
            line = line.substr(0, ignore_idx);
        }
        line = strstrip(line);
        if (line.empty()) continue;
        
        // guaranteed nonempty line
        std::vector<std::string> words = strsplit(line, " ");
        if (words.size() != 2) {
            char msg[100];
            sprintf(msg, "Invalid config file line: \"%s\".", line.c_str());
            throw std::runtime_error(msg);
        }

        words[0] = "-" + words[0];
        vargs.push_back(words[0]);
        vargs.push_back(words[1]);
    }

    // amend args object w/ command-line arg parser
    read_command_line(vargs);
}

Context::Context(int argc, char **argv) {
    // convert args to vector format
    std::vector<std::string> vargs;
    for (int i = 0; i < argc; ++i) {
        vargs.push_back(argv[i]);
    }

    // update defaults with provided values
    std::optional<std::string> cf = get_config_file(vargs);
    if (cf.has_value()) read_config_file(cf.value());
    read_command_line(vargs);

    // check whether ELF file was set
    if (!elf_file.has_value()) {
        throw std::invalid_argument("Missing ELF file argument.");
    }

    // open output file in logger
    if (output_file.has_value()) {
        logger_.open(output_file.value());
    }
}

int Context::log(const char *format, ...) {
    va_list args;
    va_start(args, format);
    int result = logger_.log(format, args);
    va_end(args);
    return result;
}

void Context::dump() {
    CacheType caches[] = {l1_cache, l2_cache, l3_cache};
    const char *assocs[] = {"DIRECT_MAPPED", "SET_ASSOC", "FULL_ASSOC"};
    const char *evicts[] = {"FIFO", "LRU", "LFU"};
    const char *coherencies[] = {"NONE", "SNOOP", "DIR"};

    log("##### CONTEXT DUMP #####\n");
    log("Config file: %s\n", config_file.has_value() ? config_file.value().c_str() : "null");
    log("ELF file: %s\n", elf_file.has_value() ? elf_file.value().c_str() : "null");
    log("Output file: %s\n", output_file.has_value() ? output_file.value().c_str() : "null");
    log("Allow RV32M: %s\n", allow_rv32m ? "true" : "false");
    log("Allow RV32A: %s\n", allow_rv32a ? "true" : "false");
    log("RAM size: %" PRIu64 "\n", ram_size);
    log("Number of harts: %d\n", num_harts);
    log("Cycle frequency: %d\n", cycle_frequency);
    log("Cache depth: %d\n", cache_depth);
    for (int i = 0; i < 3; ++i) {
        log("L%d Cache:\n", i + 1);
        log("\tAssociativity: %s\n", assocs[caches[i].ca]);
        log("\tWays: %d\n", caches[i].ways);
        log("\tSize: %" PRIu64 "\n", caches[i].size);
        log("\tBlock size: %d\n", caches[i].block_size);
        log("\tWrite policy: %s\n", caches[i].is_write_back ? "write-back" : "write-through");
        log("\tEviction policy: %s\n", evicts[caches[i].eviction]);
    }
    log("Cache coherency: %s\n", coherencies[cache_coherency]);
    log("Branch prediction rows: %d\n", branch_prediction_rows);
    log("Default preditction: %d\n", default_prediction);
    log("Allow MMU: %s\n", allow_mmu ? "true" : "false");
    log("Number of TLB slots: %d\n", num_tlb_slots);
    log("TLB eviction Policy: %s\n", evicts[tlb_eviction]);
    log("##### CONTEXT DUMP #####\n");
}

File: src/main/java/net/eabbott/riscvm/context/Context.java
package net.eabbott.riscvm.context;

import net.eabbott.riscvm.context.cache.*;
import net.eabbott.riscvm.context.elf.ELFContext;
import net.eabbott.riscvm.util.Nullable;

import java.io.FileReader;
import java.io.IOException;
import java.util.*;

/*
* The main argument parsing class.
* */
public class Context extends AbstractContext {
    // Define the valid arguments for easy argument validation
    private final String[] VALID_ARGS = {
            "-f", "-i", "-o", "-m", "-a", "-mem", "-harts", "-hz", "-cL", "-c1",
            "-c2", "-c3", "-cS1", "-cS2", "-cS3", "-cB1", "-cB2", "-cB3", "-cW1", "-cW2",
            "-cW3", "-cE1", "-cE2", "-cE3", "-cC", "-bp", "-bpD", "-mmu", "-tlb", "-tlbE"
    };
    private final Set<String> VALID_ARGS_SET = new HashSet<>(Arrays.stream(VALID_ARGS).toList());

    /*
    * A static method for creating an argument-populated context class with one function call.
    * */
    public static @Nullable Context parseArgs(String[] args) {
        Context context = new Context();
        try {
            // The main argument parsing calls
            context.configFile = context.getConfigFile(args);
            if (context.configFile != null) context.readConfigFile(context.configFile);
            context.readCommandLine(args);

            // After argument parsing, throw an exception if no ELF file was provided
            if (context.elfFile == null) {
                throw new IllegalArgumentException(
                    "Missing ELF file argument."
                );
            } else {
                context.elf = ELFContext.read(context.elfFile);
            }

            // Start a logger to print to the correct output file (or stdout)
            if (context.outputFile != null) {
                context.logger.open(context.outputFile);
            }
            return context;

        } catch (Exception e) {
            context.logger.log("Failed to parse arguments: %s", e.getMessage());
            return null;
        }
    }

    /*
    * A big, ugly function for printing all the configurations of the virtual machine.
    * */
    @Override
    public void dump() {
        CacheContext[] caches = {this.l1Cache, this.l2Cache, this.l3Cache};
        logger.log("##### CONTEXT DUMP #####");
        logger.log("Config file: %s", configFile);
        logger.log("ELF file: %s", elfFile);
        logger.log("Output file: %s", outputFile);
        logger.log("Allow RV32M: %s", allowRV32M);
        logger.log("Allow RV32A: %s", allowRV32A);
        logger.log("RAM size: %d", ramSize);
        logger.log("Number of harts: %d", numHarts);
        logger.log("Cycle frequency: %d", cycleFrequency);
        logger.log("Cache depth: %d", cacheDepth);
        for (int i = 0; i < 3; ++i) {
            logger.log("L%d Cache:", i + 1);
            logger.log("\tAssociativity: %s", caches[i].ca.type);
            logger.log("\tWays: %d", caches[i].ca.ways);
            logger.log("\tSize: %d", caches[i].size);
            logger.log("\tBlock size: %d", caches[i].blockSize);
            logger.log("\tWrite policy: %s", caches[i].isWriteBack ? "write-back" : "write-through");
            logger.log("\tEviction policy: %s", caches[i].eviction);
        }
        logger.log("Cache coherency: %s", cacheCoherency);
        logger.log("Branch prediction rows: %d", branchPredictionRows);
        logger.log("Default prediction: %d", defaultPrediction);
        logger.log("Allow MMU: %s", allowMMU);
        logger.log("Number of TLB slots: %d", numTLBSlots);
        logger.log("TLB eviction Policy: %s", tlbEviction);
        logger.log("");
        logger.log("Number of sections: %d", elf.header.phnum);
        for (var ph : elf.programHeaders) {
            logger.log("file offset = 0x%x, vaddr = 0x%x, filesz = %d, memsz = %d", ph.offset, ph.vaddr, ph.filesz, ph.memsz);
        }
        logger.log("Program counter should be set to 0x%08x", elf.header.entry);
        logger.log("##### CONTEXT DUMP #####");
    }

    /*
    * Does a quick pass through the arguments to determine if there's a config file.
    * */
    private @Nullable String getConfigFile(String[] args) throws IllegalArgumentException {
        String output = null;
        for (int i = 0; i < args.length; ++i) {
            if (args[i].equals("-f")) {
                if (i != args.length - 1) {
                    output = args[i + 1];
                    break;
                } else {
                    throw new IllegalArgumentException("Missing file name for configuration file.");
                }
            }
        }
        return output;
    }

    /*
    * Converts any valid configuration file into fake command line arguments. Forwards the fake arguments to the
    * command line argument parser.
    * */
    private void readConfigFile(String fileName) {
        // Try to open the configuration file
        try (FileReader reader = new FileReader(fileName)){
            List<String> args = new Vector<>(); // Fake arguments to pass to command line parser

            // For every line in the file...
            List<String> lines = reader.readAllLines();
            for (String line : lines) {
                int ignoreIdx = line.indexOf("#"); // ...ignore any text after a "#"
                if (ignoreIdx != -1) line = line.substring(0, ignoreIdx);
                line = line.strip();
                if (line.isEmpty()) continue;

                // ...if the line is still nonempty, compute the words
                String[] split = line.split(" ");
                List<String> words = new Vector<>();
                for (String word : split) {
                    if (!word.isBlank()) {
                        words.add(word);
                    }
                }
                if (words.size() % 2 != 0) {
                    // ...throw an exception if there are an uneven number of words in the line
                    throw new IllegalArgumentException(
                        "Argument in config file is either missing or split between lines."
                    );
                }

                // ...ensure each word is legal and add it to the fake argument list
                for (String word : words) {
                    if (word.equals("i") || word.equals("f") || word.equals("o")) {
                        throw new IllegalArgumentException(
                                "Words \"i\", \"f\", and \"o\" are illegal in configuration files."
                        );
                    }
                    args.add(args.size() % 2 == 0 ? "-" + word : word);
                }
            }

            // ...forward the fake arguments to the command line parser
            readCommandLine(args.toArray(new String[0]));

        } catch (IOException e) {
            // Throw an exception if opening the file failed.
            throw new IllegalArgumentException(
                String.format("Failed to open %s: %s", fileName, e.getMessage())
            );
        }
    }

    /*
    * The command line parsing function.
    * */
    private void readCommandLine(String[] args) {
        for (int i = 0; i < args.length; i += 2) {
            String arg = args[i];

            // Confirm argument is valid
            if (!VALID_ARGS_SET.contains(arg)) {
                throw new IllegalArgumentException(
                    String.format("Invalid argument: \"%s\".", arg)
                );
            }

            // Confirm space for additional argument
            if (i == args.length - 1) {
                throw new IllegalArgumentException(
                    String.format("Missing value for argument: \"%s\".", arg)
                );
            }
            String next = args[i + 1];

            // Unholy arg handling of despair :(
            switch (arg) {
                case "-f" -> {
                    this.configFile = next;
                }
                case "-i" -> {
                    this.elfFile = next;
                }
                case "-o" -> {
                    this.outputFile = next;
                }
                case "-m" -> {
                    this.allowRV32M = parseBoolean(next, "m", "on", "off");
                }
                case "-a" -> {
                    this.allowRV32A = parseBoolean(next, "a", "on", "off");
                }
                case "-mem" -> {
                    this.ramSize = parseMetric(next, "mem");
                }
                case "-harts" -> {
                    this.numHarts = parseInteger(next, "harts");
                }
                case "-hz" -> {
                    this.cycleFrequency = Math.toIntExact(parseMetric(next, "hz"));
                }
                case "-cL" -> {
                    this.cacheDepth = parseInteger(next, "cL");
                }
                case "-c1" -> {
                    this.l1Cache.ca = parseCacheAssociativity(next, "c1");
                }
                case "-c2" -> {
                    this.l2Cache.ca = parseCacheAssociativity(next, "c2");
                }
                case "-c3" -> {
                    this.l3Cache.ca = parseCacheAssociativity(next, "c3");
                }
                case "-cS1" -> {
                    this.l1Cache.size = parseMetric(next, "cS1");
                }
                case "-cS2" -> {
                    this.l2Cache.size = parseMetric(next, "cS2");
                }
                case "-cS3" -> {
                    this.l3Cache.size = parseMetric(next, "cS3");
                }
                case "-cB1" -> {
                    this.l1Cache.blockSize = parseInteger(next, "cB1");
                }
                case "-cB2" -> {
                    this.l2Cache.blockSize = parseInteger(next, "cB2");
                }
                case "-cB3" -> {
                    this.l3Cache.blockSize = parseInteger(next, "cB3");
                }
                case "-cW1" -> {
                    this.l1Cache.isWriteBack = parseBoolean(next, "cW1", "wb", "wt");
                }
                case "-cW2" -> {
                    this.l2Cache.isWriteBack = parseBoolean(next, "cW2", "wb", "wt");
                }
                case "-cW3" -> {
                    this.l3Cache.isWriteBack = parseBoolean(next, "cW3", "wb", "wt");
                }
                case "-cE1" -> {
                    this.l1Cache.eviction = parseEvictionPolicy(next, "cE1");
                }
                case "-cE2" -> {
                    this.l2Cache.eviction = parseEvictionPolicy(next, "cE2");
                }
                case "-cE3" -> {
                    this.l3Cache.eviction = parseEvictionPolicy(next, "cE3");
                }
                case "-cC" -> {
                    this.cacheCoherency = parseCacheCoherency(next, "cC");
                }
                case "-bp" -> {
                    this.branchPredictionRows = parseInteger(next, "bp");
                }
                case "-bpD" -> {
                    this.defaultPrediction = parseInteger(next, "bpD");
                }
                case "-mmu" -> {
                    this.allowMMU = parseBoolean(next, "mmu", "on", "off");
                }
                case "-tlb" -> {
                    this.numTLBSlots = parseInteger(next, "tlb");
                }
                case "-tlbE" -> {
                    this.tlbEviction = parseEvictionPolicy(next, "tlbE");
                }
                default -> {
                    // Should never be reached
                    throw new IllegalArgumentException(
                        String.format("Invalid argument: \"%s\".", arg)
                    );
                }
            }
        }
    }

    /*
    * Returns true if a passed argument is truthy for boolean flags.
    * */
    private boolean parseBoolean(String value, String name, String truthy, String falsey) {
        if (value.equals(truthy)) return true;
        else if (value.equals(falsey)) return false;
        throw new IllegalArgumentException(
            String.format("Argument for \"%s\" must be %s\\%s, not \"%s\".", name, truthy, falsey, value)
        );
    }

    /*
    * Returns an argument value as a parsed long, accounting for metric prefixes.
    * */
    private long parseMetric(String value, String name) {
        // Verify all but the last character are digits
        for (int i = 0; i < value.length() - 1; ++i) {
            if (!Character.isDigit(value.charAt(i))) {
                throw new IllegalArgumentException(
                    String.format("Invalid numeric for argument \"%s\": \"%s\"", name, value)
                );
            }
        }

        // Verify the last character is either a digit or a character
        char end = value.charAt(value.length() - 1);
        if (!Character.isDigit(end) && !Character.isLetter(end)) {
            throw new IllegalArgumentException(
                String.format("Invalid value for argument \"%s\": \"%s\"", name, value)
            );
        }

        // If the last character is a digit, just parse as a Long
        if (Character.isDigit(end)) {
            return Long.parseLong(value);
        }

        // Otherwise, verify value ends with a metric prefix
        if (end != 'K' && end != 'M' && end != 'G') {
            throw new IllegalArgumentException(
                String.format("Invalid metric prefix for argument \"%s\": \"%s\"", name, value)
            );
        }

        // Ensure the value isn't just the prefix
        if (value.length() == 1) {
            throw new IllegalArgumentException(
                String.format("Invalid value for argument \"%s\": \"%s\"", name, value)
            );
        }

        // Actually parse the metric value
        long numeric = Long.parseLong(value.substring(0, value.length() - 1));
        return switch (end) {
            case 'K' -> numeric * 1_024;
            case 'M' -> numeric * 1_048_576;
            case 'G' -> numeric * 1_073_741_824;
            default -> throw new IllegalArgumentException(
                String.format("Invalid value for argument \"%s\": \"%s\"", name, value)
            );
        };
    }

    /*
    * Parses an argument value as a cache associativity.
    * */
    private CacheAssociativity parseCacheAssociativity(String value, String name) {
        if (value.length() < 2) {
            throw new IllegalArgumentException(
                String.format("Invalid argument for \"%s\": \"%s\"", name, value)
            );
        }

        String start = value.substring(0, 2);
        switch (start) {
            case "dm" -> {
                return new CacheAssociativity(CacheAssociativityType.DIRECT_MAPPED, -1);
            }
            case "fa" -> {
                return new CacheAssociativity(CacheAssociativityType.FULL_ASSOC, -1);
            }
            case "sa" -> {
                if (value.length() < 4) {
                    throw new IllegalArgumentException(
                        String.format("Invalid argument for \"%s\": \"%s\"", name, value)
                    );
                }
                int ways = parseInteger(value.substring(3), name);
                return new CacheAssociativity(CacheAssociativityType.SET_ASSOC, ways);
            }
            default -> {
                throw new IllegalArgumentException(
                    String.format("Invalid argument for \"%s\": \"%s\"", name, value)
                );
            }
        }
    }

    /*
    * Returns an argument value parsed as an integer.
    * */
    private int parseInteger(String value, String name) {
        try {
            return Integer.parseInt(value);
        } catch (NumberFormatException e) {
            throw new IllegalArgumentException(
                String.format("Invalid argument for \"%s\": \"%s\"", name, value)
            );
        }
    }

    /*
    * Returns an argument value parsed as a cache coherency.
    * */
    private CacheCoherency parseCacheCoherency(String value, String name) {
        switch (value) {
            case "none" -> {
                return CacheCoherency.NONE;
            }
            case "snoop" -> {
                return CacheCoherency.SNOOP;
            }
            case "dir" -> {
                return CacheCoherency.DIR;
            }
            default -> throw new IllegalArgumentException(
                String.format("Invalid argument for \"%s\": \"%s\"", name, value)
            );
        }
    }

    /*
    * Returns an argument value parsed as an eviction policy.
    * */
    private CacheEvictionPolicy parseEvictionPolicy(String value, String name) {
        switch (value) {
            case "fifo" -> {
                return CacheEvictionPolicy.FIFO;
            }
            case "lru" -> {
                return CacheEvictionPolicy.LRU;
            }
            case "lfu" -> {
                return CacheEvictionPolicy.LFU;
            }
            default -> throw new IllegalArgumentException(
                String.format("Invalid argument for \"%s\": \"%s\"", name, value)
            );
        }
    }
}